98_spire_integration.md

```diff
@@ -0,0 +1,227 @@
+│         │         │        │          │       │                    │    Keylime    │
+│         └─────────┘        └──────────┘       │                    │   Verifier    │
+│                                               │                    │               │
+└───────────────────────────────────────────────┘                    └───────────────┘
+```
+
+This flow has the following steps:
+
+1. SPIRE Agent queries node-local /info API on keylime agent to get information like the Keylime UUID
+2. SPIRE Agent creates a nonce that is sent to the TPM’s AK (keylime created) for signing
+3. SPIRE Agent sends the information to the Spire Server
+4. SPIRE Server queries Keylime Verifier about the agent. Does it exist? Is it passing attestation? If so, can you unencrypt (verify signature) of this nonce? If all are true, then SPIRE attestation passed and identity is issued.
+
+
+In order to accomodate this flow, this enhancement will consist of the following:
+
+1. A new node-local, non-TLS API on the keylime agent responding the the `/info` path. It will return information about the keylime agent which will be used to not only identity the agent, but also be used to perform a signature verification. A 3rd party can use the credential created by the agent in the TPM to sign a nonce which can then be verified by the verifier. The new API will return the following information:
+
+    * agent_uuid
+    * tpm_hash_alg
+    * tpm_encryption_alg
+    * tpm_signing_alg
+    * ek_handle
+
+2. A new API on the verifier that can take a signed payload from a TPM and given agent's UUID verify that it came from a TPM associated with that agent. This will be used to independently verify that the Keylime agent resides on a node with that TPM.
+
+3. An expansion of the existing `/agents` GET API on the verifier to return enough information for use as selectors in SPIRE.
+
+
+### Test Plan
+
+The individual new APIs will have tests written for them. And the new
+SPIRE plugins written to use those APIs will also have their own CI/CD
+tests/pipelines to test against those APIs, targetting specific versions
+of the Keylime agent and verifier.
+
+### Upgrade / Downgrade Strategy
+
+These will be net-new APIs and will require a minor bump in the Keylime
+API version number. It's not believed that they will require database
+schema changes, nor any upgrade migrations. As such, there doesn't need
+to be a downgrade strategy.
+
+### Dependency requirements
+
+It is not believed that we will require any new dependencies for these
+APIs as they will just re-use existing libraries for any cryptographic
+signing or verification of those signatures.
+
+## Drawbacks
+
+It's possible that these APIs won't be useful outside of the SPIRE
+integration, but it's our belief they will be generic enough to be evolved
+for any 3rd party that wants to do deep verification of an node's status
+in Keylime.
+
+## Alternatives
```

Comment:
It would also be great to compare a keylime integration against the "tpm_devid" plugin here, and what the advantages for a keylime integration would be.

Reply:
Good idea, I'll add it.